// live-react-native/js/index.ts
// Main exports for LiveReact Native

// ✅ IMPLEMENTED: Core functional API
export { createLiveViewClient } from './client/LiveViewChannel';
export { useLiveView } from './hooks/useLiveView';
export { useAdvancedUpdates } from './hooks/useAdvancedUpdates';
export { usePerformanceMonitoring } from './hooks/usePerformanceMonitoring';

// ❌ STUBS: Not yet implemented - will throw errors if used
export { useLiveComponent } from './hooks/useLiveComponent'; // TODO: Phase 2.2
export { useLiveUpload } from './hooks/useLiveUpload'; // TODO: Phase 5.3
export { LiveProvider } from './components/LiveProvider'; // TODO: Phase 2.3
export { LiveComponent } from './components/LiveComponent'; // TODO: Phase 2.2
export { ComponentRegistry } from './client/ComponentRegistry'; // TODO: Phase 2.2
export { LiveViewChannel } from './client/LiveViewChannel'; // Legacy class - use createLiveViewClient instead

// Types
export type {
  LiveViewOptions,
  LiveViewState,
  PushEventFunction,
  HandleEventFunction,
  LiveComponentProps,
} from './types';

export type {
  LiveViewClientOptions,
  LiveViewClient,
} from './client/LiveViewChannel';

// Version
export const version = '0.1.0';
